warp/aggregator/aggregator.go

```diff
@@ -36,16 +71,150 @@ func (a *Aggregator) AggregateSignatures(ctx context.Context, unsignedMessage *a
+	if len(validators) == 0 {
+		return nil, fmt.Errorf("%w (SubnetID: %s, Height: %d)", errNoValidators, a.subnetID, pChainHeight)
+	}
+
+	return aggregateSignatures(ctx, a.client, unsignedMessage, validators, quorumNum, totalWeight)
+}
+
+func aggregateSignatures(
+	ctx context.Context,
+	client SignatureGetter,
+	unsignedMessage *avalancheWarp.UnsignedMessage,
+	validators []*avalancheWarp.Validator,
+	quorumNum uint64,
+	totalWeight uint64,
+) (*AggregateSignatureResult, error) {
+	type signatureFetchResult struct {
+		sig    *bls.Signature
+		index  int
+		weight uint64
+	}
+
+	// Create a child context to cancel signature fetching if we reach signature threshold.
+	signatureFetchCtx, signatureFetchCancel := context.WithCancel(ctx)
+	defer signatureFetchCancel()
+
+	// Fetch signatures from validators concurrently.
+	signatureFetchResultChan := make(chan *signatureFetchResult, len(validators))
+	for i, validator := range validators {
+		var (
+			i         = i
+			validator = validator
+			// TODO: update from a single nodeID to the original slice and use extra nodeIDs as backup.
+			nodeID = validator.NodeIDs[0]
+		)
+		go func() {
+			log.Debug("Fetching warp signature",
+				"nodeID", nodeID,
+				"index", i,
+			)
+
+			signature, err := client.GetSignature(signatureFetchCtx, nodeID, unsignedMessage)
+			if err != nil {
+				log.Debug("Failed to fetch warp signature",
+					"nodeID", nodeID,
+					"index", i,
+					"err", err,
+				)
+				signatureFetchResultChan <- nil
+				return
+			}
+
+			sigHex := hexutil.Bytes(bls.SignatureToBytes(signature))
+			log.Debug("Retrieved warp signature",
+				"nodeID", nodeID,
+				"index", i,
+				"signature", sigHex,
+			)
+
+			if !bls.Verify(validator.PublicKey, signature, unsignedMessage.Bytes()) {
+				log.Debug("Failed to verify warp signature",
+					"nodeID", nodeID,
+					"index", i,
+					"signature", sigHex,
+					"msgID", unsignedMessage.ID(),
+				)
+				signatureFetchResultChan <- nil
+				return
+			}
+
+			signatureFetchResultChan <- &signatureFetchResult{
+				sig:    signature,
+				index:  i,
+				weight: validator.Weight,
+			}
+		}()
+	}
+
+	var (
+		signatures                = make([]*bls.Signature, 0, len(validators))
+		signersBitset             = set.NewBits()
+		signaturesWeight          = uint64(0)
+		signaturesPassedThreshold = false
+	)
+
+	for i := 0; i < len(validators); i++ {
+		signatureFetchResult := <-signatureFetchResultChan
+		if signatureFetchResult == nil {
+			continue
+		}
+
+		signatures = append(signatures, signatureFetchResult.sig)
+		signersBitset.Add(signatureFetchResult.index)
+		signaturesWeight += signatureFetchResult.weight
+		log.Debug("Updated weight",
+			"totalWeight", signaturesWeight,
+			"addedWeight", signatureFetchResult.weight,
+		)
+
+		// If the signature weight meets the requested threshold, cancel signature fetching
+		if err := avalancheWarp.VerifyWeight(signaturesWeight, totalWeight, quorumNum, params.WarpQuorumDenominator); err == nil {
```

Comment:
I think we should check if err is avalancheWarp.ErrInsufficientWeight then continue else if err != nil log err or sthg so we can check unexpected errors

Reply:
`avalancheWarp.ErrInsufficientWeight` is the only error `VerifyWeight` can return so idk if we want to check for that error in particular. If people feel strongly I can change this but I don't think this would buy us anything